make 'pkg updating' to recognize regular expression package names

yuichiro-naito opened the pull request:
I added `matcher` function for matching regular expression package names in `AFFECTS:` line.
This function split `AFFECTS:` line into words and each words are checked if it contains regular expression characters or not.
If a word contains regular expression characters, it will be compiled and checked matches to ports origin name.

Compiling a regular expression is relatively slow operation, so I made a cache for compiled objects.

And I added special case of 'AFFECTS: all users' and 'AFFECTS: all ports users'.
The message in these cases should be shown.

This pull request solves [#227193](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=227193) on FreeBSD Bugzilla.

yuichiro-naito commented:
Please review and merge this pull request.

bapt commented:
thank you, I will review asap!

yuichiro-naito commented:
@bapt 
Are you busy now?

Some one crams as same as  [#227193](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=227193) in [freebsd-users-jp mailing list](https://lists.freebsd.org/pipermail/freebsd-users-jp/2019-June/001395.html).

Hope you have enough time to review.

bapt commented:
yes sorry I have been busy, I will try to review before the end of the week, sorry for the delay

bapt commented:
note that it looks globally good, (from a quick review I already did, I just want to make a more in depth review to see if I have not missed something) would it be possible for you to add some test cases?

yuichiro-naito commented:
OK, my test cases are followings.
I used the latest UPDATING file.

| Test Case | Expected Result | Comment |
|-----------|-----------------|---------|
| install net/samba48 | 20190619 is shown | includes [] expression |
| install lang/perl5.26 | 20181213 is shown | includes * expression |
| install graphics/ilmbase | 20180922 is shown | comma is just after the package name |
| (always) | 20171130 is shown | all ports users message should be always shown |
| (always) | 20131008 is shown | all users message should be always shown |
| install mysql55-server | 20150204 is shown | includes () expression |
| install postgresql95-server | 20141208 is shown | includes ? expression |
| install print/cups-base | 20110828 is shown | includes {} expression |

There is no test cases for '^', '$', '+' expressions.
If we test these cases, we have to add new "AFFECTS:" line to UPDATING file.

bapt commented:
I was more saying adding test case to the testsuite to ensure we have no regression (see the tests directory, in particular frontend/updating)

yuichiro-naito commented:
Sorry about misunderstanding.
I will try to update test code.

yuichiro-naito commented:
I added `tests/frontend/updating.sh` to test `pkg updating` message.

Through this test,
I found regular expression should be converted to POSIX style mentioned in re_format(4).
This is fixed in c7500519e31628d97b39605ee225bcf960882416.

And I think 'print/cups' package should not match to 'print/cups-base',
so I fixed to match more exactly in 827747d58986c08fbf340558056d18a033be7c5b.

yuichiro-naito commented:
@bapt 
Please check my test code, too.

bapt commented:
sure thanks a lot!

bapt commented:
Thanks a lot for the contribution!
